### ISSUE TITLE
make more apparent which ssl attributes have to be set individually

### ISSUE THREAD

**Issue by bax-** [h2o/h2o#772]
I have "default" ports and host IPs set in global context:

```
listen:
    host: &ip4 10.0.0.2
    port: 80
listen:
    host: &ip6 XX:XX:XX:XX::XX
    port: 80
listen:
    host: *ip4
    port: 443
    ssl: &SSL
      minimum-version: TLSv1.2
      dh-file: &dh-file /usr/local/etc/ssl/dhparam4096.pem
      cipher-preference: &cipher-preference server
      cipher-suite: &cipher-suite DHE-RSA-CHACHA20-POLY1305:EDH+CAMELLIA:EDH+aRSA:ECDHE-RSA-CHACHA20-POLY1305:EECDH+aRSA+AESGCM:EECDH+aRSA+SHA256:EECDH:+CAMELLIA128:+AES128:+SSLv3:!aNULL:!eNULL:!LOW:!3DES:!MD5:!EXP:!PSK:!DSS:!RC4:!SEED:!IDEA:!ECDSA:kEDH:CAMELLIA128-SHA:AES128-SHA
      certificate-file: /usr/local/etc/ssl/certs/www.example.com/fullchain.pem
      key-file:         /usr/local/etc/ssl/certs/www.example.com/privkey.pem
listen:
    host: *ip6
    port: 443
    ssl: *SSL
```

It makes little sense in my case that I have to define certificate and key to make h2o start although they will be redefined in the hosts' configs.

```
"subdomain.example.com:80":
    paths:
      "/":
        redirect:
          status: 301
          url: https://subdomain.example.com/
  "subdomain.example:443":
    header.setifempty: "strict-transport-security: max-age=15768000; includesubdomains; preload"
    access-log: /var/log/h2o/subdomain.example.com.log
    listen:
      host: *ip4
      port: 443
      ssl: &ssl_subdomain
        dh-file: *dh-file
        cipher-preference: *cipher-preference
        cipher-suite: *cipher-suite
        certificate-file: /usr/local/etc/ssl/certssubdomain.example.com/fullchain.pem
        key-file:         /usr/local/etc/ssl/certs/subdomain.example.com/privkey.pem
    listen:
      host: *ip6
      port: 443
      ssl: *ssl_subdomain
    paths:
      "/":
        proxy.reverse.url: "http://10.0.0.10/"
      "/.well-known/acme-challenge": *acme
```

I found out that if I redefine `ssl:` in a host's context (which of course I have to since I have to use different certificates and keys) that I have to re-set `dh-file`, `cipher-preference` and `cipher-suite` in the individual context again to override the defaults.

`minimum-version`, however, seems OK to be set only once in global context.

It would be nice if these details would make it into the documentation.

If I'm doing anything wrong here of if I'm misinterpreting anything, please enlighten me!

**Comment by kazuho** [comment]
> I found out that if I redefine ssl: in a host's context (which of course I have to since I have to use different certificates and keys) that I have to re-set `dh-file`, `cipher-preference` and c`ipher-suite` in the individual context again to override the defaults.

Correct.

> `minimum-version`, however, seems OK to be set only once in global context.

If it behaves that way, then it is an issue of LibreSSL (or OpenSSL, depending on how you have compiled H2O).

**Comment by bax-** [comment]
It does for me at least.
I compiled with LibreSSL 2.2.6
I have only set it in global context and all my sites only deliver TLS 1.2

By the way: This is a behavior I was expecting coming from nginx.
I think if settings from global contexts were inherited to smaller contexts like this  it would simplify and shorten h2o's configuration greatly.
Settings from global context e.g. should be taken over in hosts context if they are not redefined there.
examples: bind IP addresses, ciphers, maybe even paths (e.g. to disallow specific files globally)…

**Comment by kazuho** [comment]
> I think if settings from global contexts were inherited to smaller contexts like this it would simplify and shorten h2o's configuration greatly.
> Settings from global context e.g. should be taken over in hosts context if they are not redefined there.
examples: bind IP addresses, ciphers, maybe even paths (e.g. to disallow specific files globally)…

FWIW configuration directives do inherit the setting of outer scope (see https://h2o.examp1e.net/configure/syntax_and_structure.html#config_levels).

It is the parameter of the configuration directives (e.g. `ssl` attribute of the `listen` directive) that is not inherited.  And generally speaking they are intentionally implemented as such; for example it is necessary to require both `key-file` and `certificate-file` set for every `ssl` attribute.  If the values were inheritable, you would need to look around entire configuration file to determine if a port specified using `listen` directive is accepting either `HTTP` or `HTTPS`.

**Comment by bax-** [comment]
Thanks for pointing that out!
Though I understand the idea behind that it would make things like redirecting from port 80 to 443 for all hosts easier if it could be just set once in global context (the redirect and the other ssl settings) and just give details like `key-file` in the specific hosts.

**Comment by kazuho** [comment]
closing as a duplicate of #758.  I hope you will be satisfied by #786.

**Comment by bax-** [comment]
great, thank you!
